### ISSUE TITLE
[AI] Infra: GPU Hang / Reset (1 job)

### ISSUE THREAD

**Issue by dhonnappa-amd** [dhonnappa-amd/build-dashboards-testing#3867]
## AI Analysis — Infrastructure Issue: GPU Hang / Reset

**1 affected job**

| Pipeline/Build | Queue | Job Link |
|---------------|-------|----------|
| amd-ci/7157 | `amd_mi355_1` | [View job](https://buildkite.com/vllm/amd-ci/builds/7157#019d45ab-da8f-4d94-ae24-799513f27556) |

---

### Summary

This CI job was running vLLM MOE (Mixture of Experts) kernel tests on AMD ROCm GPUs. The job successfully completed all infrastructure setup including Docker daemon verification, image pull (27 layers), cache configuration on WekaFS, and GPU reset. The test suite began executing at 22:18:53 and ran for approximately 21 minutes, with many tests being skipped as expected. At 22:40:00, the job encountered a fatal segmentation fault (SIGSEGV) during execution of the test `test_modular_kernel_combinations_singlegpu` with specific MOE kernel parameters (dtype4, AiterExperts). The crash occurred deep in the PyTorch multiprocessing spawn code while the GPU kernel was executing, causing the Python interpreter to crash with exit code 139. The Docker container was abruptly terminated and could not be found for cleanup. This is a GPU kernel crash, not a test failure or infrastructure issue—the test code triggered a segfault in the GPU runtime or kernel code.

---

### Root Cause Analysis

Walking through the log chronologically: (1) All infrastructure setup completed successfully from 22:15:20 to 22:18:53—Docker was running, image pulled cleanly, caches configured, GPU reset performed. (2) Test execution began at 22:18:53 and progressed normally with many tests being skipped. (3) At 22:40:00.488Z, the FIRST and ONLY unrecoverable error occurred: a segmentation fault during `test_modular_kernel_combinations_singlegpu` execution. The stack trace shows the crash happened in torch.multiprocessing.spawn while executing GPU kernel code. (4) The process was killed with SIGSEGV (exit code 139), the container terminated, and the job failed. There were NO transient errors that were retried. The root cause is a GPU kernel crash—likely a memory access violation in the ROCm/HIP runtime or in the compiled Triton kernel code being tested.

---

### Detected Events (all affected jobs)

**amd-ci/7157** (queue: `amd_mi355_1`):
- `gpu_hang`: kernels/moe/test_modular_kernel_combinations.py::test_modular_kernel_combinations_singlegpu[2048-1024-32-dtype4-quant_config4-MoEPrepareAndFinalizeNoDPEPModular-AiterExperts-1] Fatal Python error: Segmentation fault

Thread 0x000075140b7af000 (most recent call first):
  File "/usr/local/lib/python3.12/dist-packages/torch/multiprocessing/spawn.py", line 140 in join
  - _The test process crashed with a segmentation fault during GPU kernel execution. The stack trace shows the crash occurred during torch multiprocessing spawn while running a MOE (Mixture of Experts) kernel test. This is a fatal GPU-related crash, not a test assertion failure._
- `gpu_hang`: /bin/bash: line 1:     8 Segmentation fault      pytest -v -s kernels/moe --ignore=kernels/moe/test_modular_oai_triton_moe.py --shard-id=$BUILDKITE_PARALLEL_JOB --num-shards=$BUILDKITE_PARALLEL_JOB_COUNT
Error response from daemon: No such container: rocm_4af9022b21aefbf366262c84196a72b6f257c753_0Ral9QCRJS
🚨 Error: The command exited with status 139
  - _Exit status 139 (128 + 11) indicates the process was killed by SIGSEGV (signal 11), confirming a segmentation fault. The container was terminated abruptly, and Docker couldn't find it for cleanup. This is the final fatal error that caused the job to fail._
- `infrastructure_success`: Docker is up. Continuing...
Setting up ModelScope cache on WekaFS
Setting up vLLM caches on WekaFS
Patching run-amd-test.sh to pass cache env vars into inner Docker container
Patched successfully
Performing GPU reset
  - _All infrastructure setup steps completed successfully. Docker daemon was running, caches were configured, and GPU reset was performed. These are NOT the cause of failure._
- `infrastructure_success`: 4af9022b21aefbf366262c84196a72b6f257c753: Pulling from rocm/vllm-ci
...
Status: Downloaded newer image for rocm/vllm-ci:4af9022b21aefbf366262c84196a72b6f257c753
docker.io/rocm/vllm-ci:4af9022b21aefbf366262c84196a72b6f257c753
  - _Docker image pull completed successfully. All 27 layers were pulled without errors. This is NOT the cause of failure._
- `test_progress`: kernels/moe/test_grouped_topk.py::test_grouped_topk[bias_dtype0-input_dtype1-1.0-softmax-True-512-22-1-1-1024-1] SKIPPED
...
kernels/moe/test_modular_kernel_combinations.py::test_modular_kernel_combinations_multigpu[2048-1024-32-dtype22-quant_config22-DeepEPLLPrepareAndFinalize-NaiveBatchedExperts-2] SKIPPED
  - _Many tests were skipped (likely due to test sharding or platform-specific conditions). Tests were running normally until the segmentation fault occurred at 22:40:00. The crash happened during test execution, not during setup._

---

#### Immediate Fix

1. Check if this specific test parameter combination (dtype4, quant_config4, MoEPrepareAndFinalizeNoDPEPModular, AiterExperts) is known to cause issues on this ROCm version. 2. Review recent commits to the MOE kernel code in commit 4af9022b21aefbf366262c84196a72b6f257c753 for memory safety issues. 3. Re-run the job to determine if this is reproducible or a one-time GPU state corruption. 4. If reproducible, skip this specific test parameter combination temporarily with pytest.mark.skip.

#### Suggested Code Change

```
In tests/kernels/moe/test_modular_kernel_combinations.py, add conditional skip:

python
@pytest.mark.skipif(
    condition=(dtype == 'dtype4' and quant_config == 'quant_config4' and 
               prepare_finalize == 'MoEPrepareAndFinalizeNoDPEPModular' and 
               experts == 'AiterExperts'),
    reason="Segfault on ROCm with this parameter combination - issue #XXXX"
)
def test_modular_kernel_combinations_singlegpu(...):


Investigate the AiterExperts implementation in vllm/model_executor/layers/fused_moe/aiter_experts.py for buffer overflows or invalid memory access patterns. Add bounds checking and validate tensor shapes before kernel launch.
```

#### Prevention

1. Add memory sanitizer runs to CI for GPU kernel tests (if ROCm supports ASAN). 2. Implement pre-test GPU memory validation to catch corruption early. 3. Add timeout wrapper around individual test cases (currently only job-level timeout exists) to prevent single test from hanging entire suite. 4. Enable core dumps for segfaults and configure automatic upload to artifact storage for post-mortem analysis. 5. Add ROCm version and GPU firmware version to test metadata to correlate crashes with specific hardware/driver combinations.

#### Workaround

Temporarily disable the failing test parameter combination using pytest markers. Run the test suite with `--ignore` flag for this specific test file until the kernel bug is fixed. Alternatively, run this test in isolation with reduced batch size or different dtype to see if the crash is parameter-dependent. Monitor GPU temperature and power during test execution to rule out hardware thermal issues.

---

_This issue was automatically created by the AI log analysis pipeline._
